train.py

```diff
@@ -8,9 +8,15 @@
 import os
 import time
 from datetime import timedelta
+from functools import partial
+
+from typing import List, Optional, Set
```

Comment:
Let's put these extra imports, including `partial`, `DeviceMesh` to the separate util file, as commented below. This can help `train.py` stay as clean as possible.

Reply:
I'm not sure if this helps keep the file clean. Suppose we put `List` in `torchtitan/utils.py`, then we need to use `List`, we instead use `utils.List` which does not look straightforward to users (this may look like a custom user-defined type).

Let me think of a better way if there's one.

Reply:
I think what @tianyu-l  means is that if we move the new logic of `get_train_context` to `parallelisms/utils.py` file, these new import will also be removed from `train.py`. Otherwise, indirect importing is not a good idea.

Reply:
I moved `get_train_context` to `utils.py`. Let me know if this is not okay. cc @fegin @tianyu-l